ckanext/dataset_subscriptions/actions/whatsapp_notifications.py

```diff
@@ -96,3 +110,77 @@ def _validate_plugin_extras(extras):
     for field in CUSTOM_FIELDS:
         out_dict[field['name']] = extras.get(field['name'], field['default'])
     return out_dict
+
+
+def get_phonenumber(user_dict):
+    if "activity_streams_whatsapp_notifications" in user_dict and "phonenumber" in user_dict:
+        if user_dict["activity_streams_whatsapp_notifications"] and user_dict["phonenumber"]:
+            phonenumber = user_dict["phonenumber"]
+            return phonenumber
+    return False
+
+
+def send_whatsapp_notifications(context, data_dict):
+    context = {'model': model, 'session': model.Session, 'ignore_auth': True}
+    users = logic.get_action('user_list')(context, {'all_fields': True})
+    for user in users:
+        user = logic.get_action('user_show')(context, {'id': user['id'],
+                                                       'include_plugin_extras': False})
+        if get_phonenumber(user):
+            return prepare_whatsapp_notifications(user)
+
+
+def prepare_whatsapp_notifications(user):
+    whatsapp_notifications_since = toolkit.config.get(
+            'ckanext.dataset_subscriptions.whatsapp_notifications_hours_since', 1)
+    whatsapp_notifications_since = int(whatsapp_notifications_since)
+    whatsapp_notifications_since = timedelta(hours=whatsapp_notifications_since)
+    whatsapp_notifications_since = (datetime.utcnow()
+                                    - whatsapp_notifications_since)
+    activity_stream_last_viewed = (
+            model.Dashboard.get(user['id']).activity_stream_last_viewed)
+    since = max(whatsapp_notifications_since, activity_stream_last_viewed)
+    return dms_whatsapp_notification_provider(user, since)
+
+
+def dms_whatsapp_notification_provider(user_dict, since):
+    context = {'model': model, 'session': model.Session,
+               'user': user_dict['id']}
+    activity_list = logic.get_action('dashboard_activity_list')(context, {})
+    dataset_activity_list = [activity for activity in activity_list
+                             if activity['user_id'] != user_dict['id']
+                             and 'package' in activity['activity_type']]
+    # We want a notification per changed dataset, not a list of all changes
+    timestamp_sorted_activity_list = sorted(dataset_activity_list,
+                                            key=lambda item: item['timestamp'])
+    deduplicated_activity_list = list({item["object_id"]:
+                                       item for item in timestamp_sorted_activity_list}.values())
+    activity_list_with_dataset_name = actions._add_dataset_name_to_activity_list(deduplicated_activity_list, context)
```

Comment:
you shouldn't be using an underscore method from another module. Those are meant to be "private" implementation details of the module.
If you found the need to reuse this code it should be moved to common module e.g. `ckanext.dataset_subscriptions.helpers` and imported from there here and in actions.